crates/storage/provider/src/providers/rocksdb/provider.rs

```diff
@@ -0,0 +1,561 @@
+    }
+
+    /// Sets a custom block cache size.
+    pub fn with_block_cache_size(mut self, capacity_bytes: usize) -> Self {
+        self.block_cache = Cache::new_lru_cache(capacity_bytes);
+        self
+    }
+
+    /// Builds the [`RocksDBProvider`].
+    pub fn build(self) -> ProviderResult<RocksDBProvider> {
+        let options =
+            Self::default_options(self.log_level, &self.block_cache, self.enable_statistics);
+
+        let cf_descriptors: Vec<ColumnFamilyDescriptor> = self
+            .column_families
+            .iter()
+            .map(|name| {
+                ColumnFamilyDescriptor::new(
+                    name.clone(),
+                    Self::default_column_family_options(&self.block_cache),
+                )
+            })
+            .collect();
+
+        let db = DB::open_cf_descriptors(&options, &self.path, cf_descriptors).map_err(|e| {
+            ProviderError::Database(DatabaseError::Open(DatabaseErrorInfo {
+                message: e.to_string().into(),
+                code: -1,
+            }))
+        })?;
+
+        let metrics = self.enable_metrics.then(RocksDBMetrics::default);
+
+        Ok(RocksDBProvider(Arc::new(RocksDBProviderInner { db, metrics })))
+    }
+}
+
+/// Some types don't support compression (eg. B256), and we don't want to be copying them to the
+/// allocated buffer when we can just use their reference.
+macro_rules! compress_to_buf_or_ref {
+    ($buf:expr, $value:expr) => {
+        if let Some(value) = $value.uncompressable_ref() {
+            Some(value)
+        } else {
+            $buf.clear();
+            $value.compress_to_buf(&mut $buf);
+            None
+        }
+    };
+}
+
+/// `RocksDB` provider for auxiliary storage layer beside main database MDBX.
+#[derive(Debug)]
+pub struct RocksDBProvider(Arc<RocksDBProviderInner>);
+
+/// Inner state for `RocksDB` provider.
+#[derive(Debug)]
+struct RocksDBProviderInner {
+    /// `RocksDB` database instance.
+    db: DB,
+    /// Metrics latency & operations.
+    metrics: Option<RocksDBMetrics>,
+}
+
+impl Clone for RocksDBProvider {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl RocksDBProvider {
+    /// Creates a new `RocksDB` provider.
+    pub fn new(path: impl AsRef<Path>) -> ProviderResult<Self> {
+        RocksDBBuilder::new(path).build()
+    }
+
+    /// Gets the column family handle for a table.
+    fn get_cf_handle<T: Table>(&self) -> Result<&rocksdb::ColumnFamily, DatabaseError> {
+        self.0
+            .db
+            .cf_handle(T::NAME)
+            .ok_or_else(|| DatabaseError::Other(format!("Column family '{}' not found", T::NAME)))
+    }
+
+    /// Executes a function and records metrics with the given operation and table name.
+    fn execute_with_operation_metric<T>(
+        &self,
+        operation: RocksDBOperation,
+        table: &'static str,
+        f: impl FnOnce(&Self) -> T,
+    ) -> T {
+        let start = std::time::Instant::now();
```

Comment:
this start is redundant if no metrics, so we can make this an optional as well and then simply do a match over both options for recording